src/System.Private.CoreLib/shared/System/Environment.cs

```diff
@@ -129,10 +129,22 @@ public static OperatingSystem OSVersion
 
         public static bool UserInteractive => true;
 
-        // Previously this represented the File version of mscorlib.dll.  Many other libraries in the framework and outside took dependencies on the first three parts of this version 
-        // remaining constant throughout 4.x.  From 4.0 to 4.5.2 this was fine since the file version only incremented the last part. Starting with 4.6 we switched to a file versioning
-        // scheme that matched the product version.  In order to preserve compatibility with existing libraries, this needs to be hard-coded.
-        public static Version Version => new Version(4, 0, 30319, 42000);
+        public static Version Version
+        {
+            get
+            {
+                // FX_PRODUCT_VERSION is expected to be set by the host
```

Comment:
It doesn't appear to be set by corerun, unless I'm just missing it. Can we do that?

Reply:
What should we set it to in corerun?

Reply:
Whatever we expect it to be set to by dotnet? It otherwise seems strange to me that the same build of coreclr would produce different results for Environment.Version. I guess my concern is that this is configurable in this way at all by the host. Why is that the approach?

Reply:
We do not have product version (ie the marketing version) available when building CoreCLR repo: https://github.com/dotnet/corefx/issues/31099#issuecomment-407190291

I guess we can redo how the build works to make it available somehow, but I have no idea what it would take.

Reply:
> I guess we can redo how the build works to make it available somehow

This makes the most sense to me, Environment.Version always returning the same version for the same CoreLib.dll build, regardless of who is hosting it, what an appcontext variable was set to, etc. It just doesn't seem like something that should be dynamic in any way.

But I also don't know what it would take. How does `dotnet` know the marketing version? Is that not configured in its build somehow?

Reply:
> Environment.Version always returning the same version for the same CoreLib.dll build

+1

> How does `dotnet` know the marketing version?

It reads it from `*.deps.json` file (either from shared framework, or from the one in self-contained app).

> I also don't know what it would take.

It seems to be available here: https://github.com/dotnet/coreclr/blob/master/eng/Versions.props#L13 so we just need to embed this property in the CoreLib.

Reply:
@mmitche Could you please confirm whether `MicrosoftNETCoreAppVersion` property is set to the current product version during the official builds?

Reply:
It wouldn't be (neither for 2.1 or 2.2), The version is flowed from the last build of core-setup that was done. Since this is a circular dependency, there's no to have pre-knowledge of what the netcore app version will be.

Reply:
Ok, I think the best compromise is to use `FX_PRODUCT_VERSION` when it is set, and use approximate version used to build CoreCLR as fallback (https://github.com/dotnet/coreclr/blob/master/eng/Versions.props#L5).

We use similar logic for `AppContext.BaseDirectory`: https://github.com/dotnet/coreclr/blob/master/src/System.Private.CoreLib/shared/System/AppContext.cs#L26

Reply:
> the best compromise

Sounds good.